posts/2022/02/making-gpus-accessible-to-pydata-ecosystem-via-array-api.md

```diff
@@ -0,0 +1,497 @@
+
+Here is an example of a naming inconsistency between NumPy and the Array API standard:
+
+```ipython
+
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Numpy API
+In [3]: np.concatenate
+Out[3]: <function numpy.concatenate>
+
+# Array API
+In [4]: npx.concat
+Out[4]: <function numpy.array_api._manipulation_functions.concat...>
+```
+
+And here is an example of a behavioral inconsistency for an indexing operation:
+
+```ipython
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Works with NumPy API
+In [3]: x = np.arange(10)
+In [4]: x[np.arange(2)]
+Out[4]: array([0, 1])
+
+
+# Doesn't Works with Array API
+In [5]: z = npx.arange(10)
+In [6]: z[npx.arange(2)]
+-----------------------------------------------------------------
+...
+IndexError: Non-zero dimensional integer array indices are not allowed in the array API namespace
+```
+
+## Results
+
+After dealing with all the challenges above, our demo works with CuPy arrays as
+input. In this section, we will talk about about its performance on GPU vs.
+CPU.
+
+We ran this on AMD GPUs as well as NVIDIA GPUs. Below we show plots comparing
+the runtime of the demo code on CPU and GPU as a function of input array size.
+The x-axis of the plot is the image dimension, i.e the dimension of the original
+image after rescaling it by a constant factor along both dimensions. For this
+demo the rescaling factors were: 0.1, 0.2, 0.4, 0.6, 0.8 and 1.
+
+
+#### On AMD GPU:
+
+![NumPy cs CuPy AMD](/images/2022/02/numpy_vs_cupy_amd.png)
+
+
+#### On NVIDIA GPU:
+
+![NumPy cs CuPy NVIDIA](/images/2022/02/numpy_vs_cupy_nvidia.png)
+
+This was run on a NVIDIA TITAN RTX. The GPU vs. CPU plot is what you would
+expect: the computation is faster with CuPy (i.e. on GPU) compared to NumPy
+(i.e. on CPU) for larger image sizes.  The computation on GPU is slower than on
+CPU for image sizes less than **61 x 77**. This is due to the overhead of
+moving data to the device (GPU), which is significant compared to the time
+taken for actual computation. The larger the image, the longer the computation
+takes, and hence the less important the data transfer becomes.
+
+
+### Reproducing these results
+
+The code and instructions to run the analysis above can be found in this repository:
+[Quansight-Labs/array-api-gpu-demo](https://github.com/Quansight-Labs/array-api-gpu-demo).
+
+The above analysis required changes in the following libraries:
+
+- `cupy`
+- `scipy`
+- `scikit-learn`
+- `scikit-image`
+
+The changes are present in the `array-api-gpu-demo` branch of [@aktech](https://github.com/aktech)'s fork
+of the above projects.
+
+In the above mentioned repo, you can find the code with which we created Docker
+images for both NVIDIA & AMD platforms. The README also contains instructions
+for how to obtain the built Docker images from Docker Registry. Once you are
+inside the Docker container, running the demo is as simple as running:
```

Comment:
I rephrased this bit to make it a little more explicit to people (like myself) who are not 100% fluent speaking Docker lingo. Please double check it all reads well @aktech